common/crypto/sanitize.go

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package crypto
+
+import (
+	"crypto/ecdsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"encoding/pem"
+	"math/big"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/hyperledger/fabric-protos-go/msp"
+	"github.com/hyperledger/fabric/bccsp/utils"
+	"github.com/pkg/errors"
+)
+
+// SanitizeIdentity sanitizes the signature scheme of the identity
+func SanitizeIdentity(identity []byte) ([]byte, error) {
+	sID := &msp.SerializedIdentity{}
+	if err := proto.Unmarshal(identity, sID); err != nil {
+		return nil, errors.Wrapf(err, "failed unmarshaling identity %s", string(identity))
+	}
+
+	der, _ := pem.Decode(sID.IdBytes)
+	if der == nil {
+		return nil, errors.Errorf("failed to PEM decode identity bytes: %s", string(sID.IdBytes))
+	}
+	cert, err := x509.ParseCertificate(der.Bytes)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed parsing certificate %s", string(sID.IdBytes))
+	}
+
+	r, s, err := utils.UnmarshalECDSASignature(cert.Signature)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed unmarshaling ECDSA signature on identity: %s", string(sID.IdBytes))
+	}
+
+	// We assume that the consenter and the CA use the same signature scheme.
+	curveOrderUsedByCryptoGen := cert.PublicKey.(*ecdsa.PublicKey).Curve.Params().N
+	halfOrder := new(big.Int).Rsh(curveOrderUsedByCryptoGen, 1)
+	// Low S, nothing to do here!
+	if s.Cmp(halfOrder) != 1 {
+		return identity, nil
+	}
+	// Else it's high-S, so shift it below half the order.
+	s.Sub(curveOrderUsedByCryptoGen, s)
+
+	var newCert certificate
+	_, err = asn1.Unmarshal(cert.Raw, &newCert)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed unmarshaling certificate")
+	}
+
+	newSig, err := utils.MarshalECDSASignature(r, s)
```

Comment:
we have this code in other places in Fabric, we should probably use that code instead of re-writing it. Let's proceed but open an issue.

Reply:
https://github.com/hyperledger/fabric/issues/4007 is updated to track this problem too.